lib/ansible/modules/cloud/amazon/iam_user.py

```diff
@@ -227,38 +227,81 @@ def create_or_update_user(connection, module):
 
 def destroy_user(connection, module):
 
-    params = dict()
-    params['UserName'] = module.params.get('name')
+    user_name = module.params.get('name')
 
-    if get_user(connection, module, params['UserName']):
-        # Check mode means we would remove this user
-        if module.check_mode:
-            module.exit_json(changed=True)
+    user = get_user(connection, module, user_name)
+    # User is not present
+    if not user:
+        module.exit_json(changed=False)
 
-        # Remove any attached policies otherwise deletion fails
-        try:
-            for policy in get_attached_policy_list(connection, module, params['UserName']):
-                connection.detach_user_policy(UserName=params['UserName'], PolicyArn=policy['PolicyArn'])
-        except ClientError as e:
-            module.fail_json(msg="Unable to detach policy {0} from user {1}: {2}".format(
-                             policy['PolicyArn'], params['UserName'], to_native(e)),
-                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
-        except ParamValidationError as e:
-            module.fail_json(msg="Unable to detach policy {0} from user {1}: {2}".format(
-                             policy['PolicyArn'], params['UserName'], to_native(e)),
-                             exception=traceback.format_exc())
+    # Check mode means we would remove this user
+    if module.check_mode:
+        module.exit_json(changed=True)
 
-        try:
-            connection.delete_user(**params)
-        except ClientError as e:
-            module.fail_json(msg="Unable to delete user {0}: {1}".format(params['UserName'], to_native(e)),
-                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
-        except ParamValidationError as e:
-            module.fail_json(msg="Unable to delete user {0}: {1}".format(params['UserName'], to_native(e)),
-                             exception=traceback.format_exc())
+    # Remove any attached policies otherwise deletion fails
+    try:
+        for policy in get_attached_policy_list(connection, module, user_name):
+            connection.detach_user_policy(UserName=user_name, PolicyArn=policy['PolicyArn'])
+    except ClientError as e:
+        module.fail_json(msg="Unable to detach policy {0} from user {1}: {2}".format(
+                         policy['PolicyArn'], user_name, to_native(e)),
+                         exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+    except ParamValidationError as e:
```

Comment:
I see you didn't change this from before, but please update this to handle BotoCoreError instead here, as ParamValidationError is only a subset of the things that could go really wrong. You'll need to import it too.

Since you ported this to AnsibleAWSModule it's preferable to use fail_json_aws and handle these two exceptions together; ClientError is only handled separately if you need to parse the .response['Error']['Code'] because it doesn't exist for BotoCoreError. `fail_json_aws` cuts down on boilerplate (it will add the traceback, to_native(e), and e.response for you if these things exist and handle it gracefully if not).
```suggestion
    except (ClientError, BotoCoreError) as e:
        module.fail_json_aws(msg="Unable to detach policy {0} from user {1}".format(policy['PolicyArn'], user_name))
```